[C++ verification] no matching constructor for initialization of 'std::vector<int>'

````CPP
#include <iostream>
#include <vector>
#include <algorithm> // for std::copy
#include <cassert>   // for assert

int main() {
    // Define a source vector of integers
    std::vector<int> source = {1, 2, 3, 4, 5};

    // Create a destination vector with the same size as the source
    std::vector<int> destination(source.size());

    // Use std::copy to copy elements from source to destination
    std::copy(source.begin(), source.end(), destination.begin());

    // Use assertions to verify that the elements were copied correctly
    assert(source.size() == destination.size() && "Source and destination sizes should be the same");

    // Check that each element in the destination matches the source
    for (size_t i = 0; i < source.size(); ++i) {
        assert(source[i] == destination[i] && "Elements in source and destination should match");
    }

    // If all assertions pass, print a success message
    std::cout << "All assertions passed! Elements were copied correctly using std::copy." << std::endl;

    return 0;
}
````

$ esbmc copy.cpp

````
Parsing copy.cpp
copy.cpp:8:22: error: no matching constructor for initialization of 'std::vector<int>'
    std::vector<int> source = {1, 2, 3, 4, 5};
                     ^        ~~~~~~~~~~~~~~~
/tmp/esbmc-cpp-headers-7404-ce26-a659/vector:134:12: note: candidate constructor not viable: requires 2 arguments, but 5 were provided
  explicit vector(iterator t1, iterator t2) : _size(0)
           ^
/tmp/esbmc-cpp-headers-7404-ce26-a659/vector:163:3: note: candidate constructor template not viable: requires 2 arguments, but 5 were provided
  vector(InputIt t, InputIt addr) : _size(0), _capacity(1)
  ^
/tmp/esbmc-cpp-headers-7404-ce26-a659/vector:177:12: note: candidate constructor not viable: requires 2 arguments, but 5 were provided
  explicit vector(size_type n, const T &x)
           ^
/tmp/esbmc-cpp-headers-7404-ce26-a659/vector:151:12: note: candidate constructor not viable: requires single argument 'n', but 5 arguments were provided
  explicit vector(size_type n) : _size(0), _capacity(1)
           ^
/tmp/esbmc-cpp-headers-7404-ce26-a659/vector:186:3: note: candidate constructor not viable: requires single argument 'x', but 5 arguments were provided
  vector(const vector &x) : _size(0), _capacity(1)
  ^
/tmp/esbmc-cpp-headers-7404-ce26-a659/vector:127:12: note: candidate constructor not viable: requires 0 arguments, but 5 were provided
  explicit vector() : _size(0), _capacity(1)
           ^
ERROR: PARSING ERROR
````

--- a/regression/esbmc-cpp/cpp/ch11_2/test.desc
+++ b/regression/esbmc-cpp/cpp/ch11_2/test.desc
@@ -1,4 +1,4 @@
-KNOWNBUG
+CORE
 main.cpp
 --unwind 10 --no-unwinding-assertions --timeout 900
 ^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp/cpp/ch11_3/test.desc
+++ b/regression/esbmc-cpp/cpp/ch11_3/test.desc
@@ -1,4 +1,4 @@
-KNOWNBUG
+CORE
 main.cpp
 --unwind 10 --no-unwinding-assertions --timeout 900
 ^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp/cpp/ch22_2/test.desc
+++ b/regression/esbmc-cpp/cpp/ch22_2/test.desc
@@ -1,4 +1,4 @@
-KNOWNBUG
+CORE
 main.cpp
 Array.cpp --unwind 10 --no-unwinding-assertions --timeout 900
 ^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp/cpp/ch7_9/test.desc
+++ b/regression/esbmc-cpp/cpp/ch7_9/test.desc
@@ -1,4 +1,4 @@
-KNOWNBUG
+CORE
 main.cpp
 employee2.cpp --unwind 10 --no-unwinding-assertions --timeout 900
 ^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp/try_catch/nec_ex7-ctor-throw/test.desc
+++ b/regression/esbmc-cpp/try_catch/nec_ex7-ctor-throw/test.desc
@@ -1,4 +1,4 @@
-KNOWNBUG
+CORE
 main.cpp
 --unwind 1 --no-unwinding-assertions --error-label ERROR --timeout 900
 

--- a/regression/esbmc-cpp/try_catch/nec_ex8-virtual-throw/test.desc
+++ b/regression/esbmc-cpp/try_catch/nec_ex8-virtual-throw/test.desc
@@ -1,4 +1,4 @@
-KNOWNBUG
+CORE
 main.cpp
 --unwind 1 --no-unwinding-assertions --error-label ERROR --timeout 900
 

--- a/regression/esbmc-cpp/vector/vector8/test.desc
+++ b/regression/esbmc-cpp/vector/vector8/test.desc
@@ -1,4 +1,4 @@
 CORE
 main.cpp
 --unwind 10 --no-unwinding-assertions --timeout 900
-^ERROR: PARSING ERROR$
+^VERIFICATION SUCCESSFUL$

--- a/src/clang-cpp-frontend/clang_cpp_adjust_expr.cpp
+++ b/src/clang-cpp-frontend/clang_cpp_adjust_expr.cpp
@@ -95,6 +95,7 @@ void clang_cpp_adjust::adjust_side_effect(side_effect_exprt &expr)
       destructor.arguments().push_back(address_of_exprt(new_object));
       expr.set("destructor", destructor);
     }
+    adjust_operands(expr);
   }
   else if (statement == "temporary_object")
   {

--- a/src/clang-cpp-frontend/clang_cpp_adjust_expr.cpp
+++ b/src/clang-cpp-frontend/clang_cpp_adjust_expr.cpp
@@ -83,6 +83,7 @@ void clang_cpp_adjust::adjust_side_effect(side_effect_exprt &expr)
   }
   else if (statement == "cpp_delete" || statement == "cpp_delete[]")
   {
+    adjust_operands(expr);
     // adjust side effect node to explicitly call class destructor
     // e.g. the adjustment here will add the following instruction in GOTO:
     // FUNCTION_CALL:  ~t2(&(*p))
@@ -95,7 +96,6 @@ void clang_cpp_adjust::adjust_side_effect(side_effect_exprt &expr)
       destructor.arguments().push_back(address_of_exprt(new_object));
       expr.set("destructor", destructor);
     }
-    adjust_operands(expr);
   }
   else if (statement == "temporary_object")
   {

--- a/src/clang-cpp-frontend/clang_cpp_convert.cpp
+++ b/src/clang-cpp-frontend/clang_cpp_convert.cpp
@@ -1012,6 +1012,36 @@ bool clang_cpp_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
     break;
   }
 
+  case clang::Stmt::CXXStdInitializerListExprClass:
+  {
+    const clang::CXXStdInitializerListExpr &cxxstdinit =
+      static_cast<const clang::CXXStdInitializerListExpr &>(stmt);
+
+    exprt list;
+    if (get_expr(*cxxstdinit.getSubExpr(), list))
+      return true;
+
+    exprt size = static_cast<const exprt &>(list.type().subtype().size_irep());
+
+    typet t;
+    if (get_type(*cxxstdinit.getType(), t))
+      return true;
+
+    const struct_union_typet &this_type = to_struct_union_type(ns.follow(t));
+    exprt sym("struct", this_type);
+
+    // { ._M_array=&list[0], ._M_len=size }
+    sym.move_to_operands(list);
+    sym.move_to_operands(size);
+
+    // Implicit construction of a std::initializer_list<T> object
+    // from an array temporary within list-initialization
+    // Therefore the AST does not call the constructor
+    new_expr = sym;
+
+    break;
+  }
+
   default:
     if (clang_c_convertert::get_expr(stmt, new_expr))
       return true;

--- a/src/cpp/library/vector
+++ b/src/cpp/library/vector
@@ -5,10 +5,12 @@
 #include <assert.h>
 
 #include "stdexcept"
-#include <initializer_list>
 
 #include <utility> /* std::move */
 #include <new> /* placement-new */
+#if __cplusplus >= 201103L
+# include <initializer_list>
+#endif
 
 #define VECTOR_CAPACITY 500
 
@@ -194,6 +196,7 @@ public:
     _capacity = _size;
   }
 
+#if __cplusplus >= 201103L
   vector(std::initializer_list<T> init) :_size(0), _capacity(1)
   {
     buf = new T[init.size()];
@@ -202,6 +205,7 @@ public:
       buf[i] = *(init.begin() + i);
     _capacity = _size;
   }
+#endif
 
   vector &operator=(const vector &x)
   {

--- a/src/cpp/library/vector
+++ b/src/cpp/library/vector
@@ -5,6 +5,8 @@
 #include <assert.h>
 
 #include "stdexcept"
+#include <initializer_list>
+
 #include <utility> /* std::move */
 #include <new> /* placement-new */
 
@@ -189,6 +191,15 @@ public:
     _size = x._size;
     for (int i = 0; i < _size; i++)
       buf[i] = x.buf[i];
+    _capacity = _size;
+  }
+
+  vector(std::initializer_list<T> init) :_size(0), _capacity(1)
+  {
+    buf = new T[init.size()];
+    _size = init.size();
+    for (int i = 0; i < _size; i++)
+      buf[i] = *(init.begin() + i);
     _capacity = _size;
   }
 

